File: clients/mobile/src-tauri/src/commands/attachments.rs
//! 附件命令（施工单 M80-03，F-09-07 / F-09-09）：上传照片或视频、取回原件。
//!
//! # 字节走原始 IPC，不走 JSON 数组
//!
//! 一段 40 MB 的视频按 `number[]` 序列化是 200 MB 的 JSON 文本，WebView 会卡几秒；
//! Tauri 2 的 raw IPC（`invoke(cmd, Uint8Array, { headers })`）把它当二进制直接递过来，
//! 元数据放请求头。Android 不支持 raw body（tauri `InvokeBody` 文档明写），那边退回
//! JSON 体里的 `bytesBase64`——两种形状这里都收。
//!
//! # 令牌不进 WebView
//!
//! 与 `chat.rs` 同一条纪律：网络与鉴权在 Rust。取件也是命令，字节以 `ipc::Response`
//! 原样回去，WebView 自己包成 blob URL 给 `<img>` / `<video>`。

use base64::Engine;
use carlife_net::{GatewayClient, UploadedAttachment};
use tauri::ipc::{InvokeBody, Request, Response};

use super::chat::gateway_env;

fn header<'a>(req: &'a Request<'_>, name: &str) -> Option<&'a str> {
    req.headers().get(name).and_then(|v| v.to_str().ok()).filter(|s| !s.is_empty())
}

/// percent-decode（`%XX` → 字节，再按 UTF-8 组回字符串）。
///
/// 解不动就**原样返回**：文件名只用于展示，不该因为一个名字编码坏了就把用户拍的照片拒掉
/// （与网关 `decodeFilename` 同一条纪律，F-09-05 边界）。
fn percent_decode(raw: &str) -> String {
    fn hex(b: u8) -> Option<u8> {
        match b {
            b'0'..=b'9' => Some(b - b'0'),
            b'a'..=b'f' => Some(b - b'a' + 10),
            b'A'..=b'F' => Some(b - b'A' + 10),
            _ => None,
        }
    }
    let bytes = raw.as_bytes();
    let mut out: Vec<u8> = Vec::with_capacity(bytes.len());
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i] == b'%' && i + 2 < bytes.len() {
            if let (Some(h), Some(l)) = (hex(bytes[i + 1]), hex(bytes[i + 2])) {
                out.push(h * 16 + l);
                i += 3;
                continue;
            }
        }
        out.push(bytes[i]);
        i += 1;
    }
    String::from_utf8(out).unwrap_or_else(|_| raw.to_string())
}

/// 上传一个附件到当前会话。请求头：`x-session-id`（必填）、`content-type`、
/// `x-filename`（**percent-encoded 的原始文件名**，见下）、`x-idempotency-key`。
/// 返回网关回执（句柄 / 类别 / 字节数）。
///
/// # 文件名在这一跳是编码过的（M80-04）
///
/// WebView 那侧必须先编码——请求头是 ByteString，「特斯拉.png」会让 `new Headers` 当场抛
/// `TypeError: Type error`（2026-09-09 真机上就是这么失败的，INC-0130）。这里解回真名，
/// 再交给 `upload_attachment`，由它按网关的约定重新编码。**每一跳各自负责自己的编码**。
#[tauri::command]
pub async fn upload_attachment(request: Request<'_>) -> Result<UploadedAttachment, String> {
    let session_id = header(&request, "x-session-id")
        .ok_or_else(|| "缺少 x-session-id".to_string())?
        .to_string();
    let content_type = header(&request, "content-type")
        .unwrap_or("application/octet-stream")
        .to_string();
    let filename = header(&request, "x-filename").map(percent_decode);
    let idempotency_key = header(&request, "x-idempotency-key").map(|s| s.to_string());
    let bytes: Vec<u8> = match request.body() {
        InvokeBody::Raw(b) => b.clone(),
        InvokeBody::Json(v) => {
            let b64 = v
                .get("bytesBase64")
                .and_then(|x| x.as_str())
                .ok_or_else(|| "JSON 体缺少 bytesBase64".to_string())?;
            base64::engine::general_purpose::STANDARD
                .decode(b64)
                .map_err(|e| format!("bytesBase64 不是合法 base64：{e}"))?
        }
    };
    if bytes.is_empty() {
        return Err("文件是空的".into());
    }
    let (base_url, token) = gateway_env();
    GatewayClient::new(base_url, token)
        .upload_attachment(
            &session_id,
            &content_type,
            filename.as_deref(),
            idempotency_key.as_deref(),
            bytes,
        )
        .await
        .map_err(|e| e.to_string())
}

/// 取回附件原件，字节原样回 WebView（`ArrayBuffer`）。MIME 由调用方从 `AttachmentRef.contentType` 取。
#[tauri::command]
pub async fn fetch_attachment(handle: String) -> Result<Response, String> {
    let (base_url, token) = gateway_env();
    let (_content_type, bytes) = GatewayClient::new(base_url, token)
        .fetch_attachment(&handle)
        .await
        .map_err(|e| e.to_string())?;
    Ok(Response::new(bytes))
}

#[cfg(test)]
mod tests {
    use super::percent_decode;

    #[test]
    fn 解码中文文件名() {
        assert_eq!(percent_decode("%E7%89%B9%E6%96%AF%E6%8B%89.png"), "特斯拉.png");
    }

    #[test]
    fn 纯_ascii_原样返回() {
        assert_eq!(percent_decode("IMG_0001.HEIC"), "IMG_0001.HEIC");
    }

    #[test]
    fn 坏编码不吞文件名() {
        // 半截的 %E7、非法十六进制、结尾孤零零一个 % —— 都不该让上传失败
        assert_eq!(percent_decode("a%zz.png"), "a%zz.png");
        assert_eq!(percent_decode("tail%"), "tail%");
        assert_eq!(percent_decode("%E7%89"), "%E7%89");
    }

    #[test]
    fn 空格与加号不被当成编码字符() {
        // encodeURIComponent 把空格编成 %20；`+` 是原样的字面量，不该被解成空格
        assert_eq!(percent_decode("my%20photo+1.jpg"), "my photo+1.jpg");
    }
}

File: clients/shared/ui/src/web-shim/attachments.ts
/**
 * 垫片的附件通道（ACR-049 追加）：上传照片 / 视频、取回原件。
 *
 * # 为什么原先是"拒绝"、现在能做
 *
 * 原生端的字节走 Tauri 的原始 IPC（`invoke(cmd, Uint8Array, { headers })`）进 Rust，
 * 再由 Rust 发给网关。当初把它列进拒绝清单的理由是"走 Rust 的分片通道"——但网关那一跳
 * 根本不分片：`POST /v1/session/:id/attachments` 就是**原始 body + 头部元数据**
 * （`gateway/src/upload/index.ts`），浏览器的 `fetch` 直接能发。
 *
 * # 两个不报错的坑
 *
 *  1. Tauri 的 `mockIPC` **丢掉 `invoke` 的第三个参数**（`mocks.js`：`invoke(cmd, args, _options)`
 *     只回调 `cb(cmd, args)`）。会话号、MIME、文件名全在那个参数的 `headers` 里——
 *     所以 `installWebShim` 在 mockIPC 之上再包一层，把 options 递进来（见 index.ts）。
 *  2. 原始 IPC 的 `args` 是 `Uint8Array` 不是对象。垫片原先把"非对象形参"一律当空参数，
 *     字节在进门那一刻就没了。
 *
 * # 头值原样转发
 *
 * 端上的 `buildUploadHeaders` 已经把 `x-filename` 按 `encodeURIComponent` 编好，
 * 这正是网关 `decodeFilename` 的约定。原生端多出的"Rust 解码再编码"那一跳在这里不存在，
 * 所以不解不编，原样递过去；幂等键同理（Rust 那侧也是原样转发）。
 */

import { GatewayError, type Gateway } from "./gateway";

/** `invoke` 第三个参数里垫片关心的部分。形状照 `@tauri-apps/api/core` 的 `InvokeOptions`。 */
export interface InvokeOptionsLike {
  headers?: Headers | Record<string, string> | [string, string][];
}

/** 要带着 options 才能工作的命令。只有它们绕过 mockIPC 的 `cb(cmd, args)`。 */
export const RAW_IPC_COMMANDS: ReadonlySet<string> = new Set(["upload_attachment"]);

/** 原始 IPC 的请求体：`Uint8Array` / `ArrayBuffer` / `number[]`（Tauri 的 `InvokeArgs` 三种都收）。 */
export function bytesOf(raw: unknown): Uint8Array | null {
  if (raw instanceof Uint8Array) return raw;
  if (raw instanceof ArrayBuffer) return new Uint8Array(raw);
  if (ArrayBuffer.isView(raw)) return new Uint8Array(raw.buffer, raw.byteOffset, raw.byteLength);
  if (Array.isArray(raw) && raw.every((n) => typeof n === "number")) return Uint8Array.from(raw as number[]);
  return null;
}

/**
 * 把网关的拒绝翻成界面能直接显示的一句话。
 *
 * 抛**裸字符串**：Rust 命令的 `Err(String)` 到 JS 这边就是字符串，界面按 `String(err)` 显示。
 * 网关的 413 / 415 带面向用户的 `reason`（"文件太大…"），优先用它；
 * 413 还可能来自 nginx（请求体上限），那时 body 是一页 HTML，没有 reason。
 */
export function uploadFailureText(err: unknown): string {
  if (!(err instanceof GatewayError)) return err instanceof Error ? err.message : String(err);
  try {
    const parsed = JSON.parse(err.body) as { reason?: unknown; error?: unknown };
    if (typeof parsed.reason === "string" && parsed.reason) return parsed.reason;
    if (parsed.error === "attachments_unavailable") return "服务端没有接对象存储，暂时收不了附件";
    if (typeof parsed.error === "string" && parsed.error) return "上传被拒绝（" + parsed.error + "）";
  } catch {
    // body 不是 JSON（nginx 的错误页）——落到下面按状态码说
  }
  if (err.status === 413) return "文件太大，传输被中断了。";
  if (err.status === 429) return "操作太频繁，稍等几秒再试。";
  return "上传失败（" + err.status + "）";
}

export interface UploadedAttachment {
  handle: string;
  kind: string;
  bytes: number;
}

/** `upload_attachment`：返回形状与 Rust 的 `UploadedAttachment` 一致（句柄 / 类别 / 字节数）。 */
export async function uploadAttachment(gw: Gateway, rawArgs: unknown, options: InvokeOptionsLike | undefined): Promise<UploadedAttachment> {
  const headers = new Headers(options?.headers);
  const sessionId = headers.get("x-session-id");
  // 与 Rust 命令同一组原因串
  if (!sessionId) throw "缺少 x-session-id";
  const bytes = bytesOf(rawArgs);
  if (!bytes || bytes.byteLength === 0) throw "文件是空的";

  const forward: Record<string, string> = { "content-type": headers.get("content-type") || "application/octet-stream" };
  for (const name of ["x-filename", "x-idempotency-key"]) {
    const value = headers.get(name);
    if (value) forward[name] = value;
  }
  try {
    const res = await gw.request("POST", "/v1/session/" + encodeURIComponent(sessionId) + "/attachments", {
      body: bytes as unknown as BodyInit,
      headers: forward,
    });
    const r = (await res.json()) as UploadedAttachment;
    return { handle: r.handle, kind: r.kind, bytes: r.bytes };
  } catch (err) {
    throw uploadFailureText(err);
  }
}

/**
 * `fetch_attachment`：字节原样回界面（`ArrayBuffer`），界面自己包成 blob URL。
 * 不能让 `<img src>` 直连 `/v1/attachments/:handle`——access token 只有 15 分钟，
 * 而 401 重登只在 `gw.request` 这条路上。
 */
export async function fetchAttachment(gw: Gateway, handle: unknown): Promise<ArrayBuffer> {
  if (typeof handle !== "string" || !handle) throw "缺少 handle";
  try {
    return await (await gw.request("GET", "/v1/attachments/" + encodeURIComponent(handle))).arrayBuffer();
  } catch (err) {
    throw err instanceof GatewayError ? "取件失败（" + err.status + "）" : String(err);
  }
}
